### ISSUE TITLE
Unsupported HLS KEYFORMAT com.apple.streamkingkeydelivery

### ISSUE THREAD

**Issue by alexanderturinske** [google/shaka-player#2096]
Releates to: #896

**Comment by TheModMaker** [comment]
We only support FairPlay through native `src=` playback, we don't support it through MSE playback.  If you don't pass the third argument to `load`, we'll default to using `src=` and it should work fine.  You should also look at the new [FairPlay tutorial](https://nightly-dot-shaka-player-demo.appspot.com/docs/api/tutorial-fairplay.html) if you are having any trouble setting up `src=` playback.

**Comment by alexanderturinske** [comment]
Thanks!

Unfortunately, that gets me back to the error I was hoping to fix by switching to shaka-player 🤦‍
```
{
  category: 6
  code: 6006
  data: Array (3)
    0 "EME PatchedMediaKeysApple key error"
    1 Error: EME PatchedMediaKeysApple key error
      column: 162
      errorCode: WebKitMediaKeyError {code: 1, systemCode: 0, MEDIA_KEYERR_UNKNOWN: 1, ... }
      line: 210074
      message: "EME PatchedMediaKeysApple key error"
      sourceURL: "https://dev.pixcf.com:3000/256404836e07da8a76d4.js"
      stack: "onWebkitKeyError_@https://dev.pixcf.com:3000
  2 undefined
handled: false
message: "Shaka Error DRM.FAILED_TO_GENERATE_LICENSE_REQUEST (EME PatchedMediaKeysApple key error,Error: EME PatchedMediaKeysApple key error,)"
severity: 2
stack: ...
}
```
Switching to shaka-player was not completely a waste though, the better error handling at least gave me some more information about the issue. I knew it was an `EME` issue, but now I have a few more things to search for.

From the docs
```

FAILED_TO_GENERATE_LICENSE_REQUEST | 6006 | number | The CDM was unable to generate a license request for the init data it was given.  The init data may be malformed or in an unsupported format.  error.data[0] is an error message string from the browser.  error.data[1] is the error object from the browser.  error.data[2] is a string with the extended error code, if available.
```

Any thoughts on this one?

**Comment by TheModMaker** [comment]
The server certificate needs to be a `Uint8Array`.  So if you are using `XMLHttpRequest` with an `arraybuffer` type or using `fetch`, then the result is an `ArrayBuffer`.  You may need to wrap the result in a `Uint8Array`:

```js
const req = await fetch('https://example.com/cert.der');
const cert = await req.arrayBuffer();

player.configure({
  drm: {
    advanced: {
      'com.apple.fps.1_0': {serverCertificate: new Uint8Array(cert)},
    },
  }
});
```

**Comment by alexanderturinske** [comment]
Thanks for the suggestion; I've got that going on.
```
player.configure({
  drm: {
    servers: {
      'com.widevine.alpha': `${this.baseUrl}/items/${id}/drm/widevine`,
      'com.microsoft.playready': `${this.baseUrl}/items/${id}/drm/playready`,
      'com.apple.fps.1_0': `${this.baseUrl}/items${id}/drm/fairplay`,
    },
    advanced: {
      'com.apple.fps.1_0': {
        serverCertificate: fetchCertificate(),
      },
    },
  },
});
...
fetchCertificate() {
  const cachedCert = certificateCache.get(this) && certificateCache.get(this).cert;
  if (cachedCert) {
    console.log('cachedCert: ', cachedCert);
    return cachedCert;
  } else {
    return axios({
      method: 'get',
      url: this.serverCertificatePath,
      responseType: 'arraybuffer',
      headers: this.headers,
    })
    .then(response => {
      const cert = new Uint8Array(response.data);
      certificateCache.set(this, { cert });
      return cert;
    })
    .catch(err => {
      console.log('fetchCertificate: err: ', err);
    })
  }
}
```
And I have verified that the cert is coming back as an array buffer.

**Comment by TheModMaker** [comment]
Your `fetchCertificate` method returns a `Promise`.  You need to wait for it to complete.  For example:

```js
player.configure({
  drm: {
    advanced: {
      'com.apple.fps.1_0': {
        serverCertificate: await fetchCertificate(),
      },
    },
  },
});
// Or
fetchCertificate().then((cert) => {
  player.configure({
    drm: {
      advanced: {
        'com.apple.fps.1_0': {
          serverCertificate: cert,
        },
      },
    },
  });
});
```

**Comment by alexanderturinske** [comment]
Yup, you are absolutely correct. After fixing that, some errors went away, but ultimately I ended up with the same error.

I continued debugging and saw that our DRM license request was failing, and found out that that the url needed to go from `${this.baseUrl}/items${id}/drm/fairplay` to `${this.baseUrl}/items${id}/drm/fairplay/license`.

Now the DRM is receiving the response, but it giving me an error. 😑 

In any case, the `Shaka Error DRM.FAILED_TO_GENERATE_LICENSE_REQUEST` error message really helped me realize that this is all a licensing request issue and that has been great.

Thanks for your help. I'll keep you posted on my progress.

**Comment by TheModMaker** [comment]
If you are getting `DRM.LICENSE_RESPONSE_REJECTED`, then one likely problem is your license server is wrapping the response.  You may need to unwrap the response before it is passed to the browser.  You can read more about that in our [tutorial](https://shaka-player-demo.appspot.com/docs/api/tutorial-license-wrapping.html).

**Comment by alexanderturinske** [comment]
Sorry for the late response; I got pulled into some fires.

Nah, I am still getting the same error (`Shaka Error DRM.FAILED_TO_GENERATE_LICENSE_REQUEST (EME PatchedMediaKeysApple key error,Error: EME PatchedMediaKeysApple key error,`), and now our license server is erroring with  
`FPSUserException: SPC byte array too short! Cannot contain payload!`, so I will look through your tutorial and see where I messed up.

**Comment by alexanderturinske** [comment]
Thanks for the tutorial. 

In our original implementation, the other library we were using just gave us the `WebKitMediaKeyMessageEvent.message`, which is a Uint8Array, and then we converted it to base64Encoded and sent it to our license server.

In looking into [the shaka-player code](https://github.com/google/shaka-player/blob/c5b9d6804f0cd351c673e9a12385e92bdd406da1/lib/polyfill/patchedmediakeys_apple.js#L574), I see that shaka-player does not simply append the `WebKitMediaKeyMessageEvent.message` to the `request.body`. 

I implemented a request filter and was able to take a look at what the request body looked like and I see that is not a Uint8Array and if I convert it to one, it doesn't match what I am seeing on the `WebKitMediaKeyMessageEvent.message` from the debugger.

Looking at the [Fairplay tutorial](https://github.com/google/shaka-player/blob/c5b9d6804f0cd351c673e9a12385e92bdd406da1/docs/tutorials/fairplay.md), I converted the `request.body` to what I think was the original Uint8Array via the following
```
const string = StringUtils.fromUTF8(request.body); //-> "spc=AAA..."
const sub = string.substring(4); //-> "AAA..."
const uint8array = Uint8ArrayUtils.fromBase64(sub); //-> Uint8 [ 0 ....]
```
, but got the same error from my server.

So, how can I take the `request.body` and convert it back to the Uint8Array I see on `WebKitMediaKeyMessageEvent.message`?

**Comment by TheModMaker** [comment]
> Looking at the Fairplay tutorial, I saw that in the example the filter was adding some additional information to the request.body, so fiddling with that, I found that the request.body has the extra spc= prepended to it, so I converted the request.body to the original Uint8Array via the following

That is because you are probably copying the FairPlay tutorial, which has a call to `player.getNetworkingEngine().registerRequestFilter()` which converts the body to that URL.  You could just convert that filter to do what you want.

> Actually, I can get playback on the first clip I play, the first time I play it, but I still get the error
> `category: 6, code: 6006, WebKitMediaKeyError {code: 1, systemCode: 0}, message: "EME PatchedMediaKeysApple key error"`

This may be #2031.  You could try changing the [error listener](https://github.com/google/shaka-player/blob/1ea3ca49dcb17d620395327314ac327a883c206b/lib/polyfill/patchedmediakeys_apple.js#L610) to ignore the error and see if you get playback.

> before the response filter is hit. The only errors I see are way before this one
> `[Error] Invalid config, unrecognized key .streaming.abr`
> `[Error] Invalid config, unrecognized key .streaming.drm`
> Could those attribute to this error?

Those are warnings from Shaka Player about invalid configuration paths.  The `abr` configuration should be at the top-level, not part of `streaming`:

```js
player.configure({
  abr: ...
});
```

**Comment by alexanderturinske** [comment]
Ah! #2031 is the github issue I have been looking for this entire time. I am converting from videojs-contrib-eme because of this issue.

> That is because you are probably copying the FairPlay tutorial, which has a call to player.getNetworkingEngine().registerRequestFilter() which converts the body to that URL. You could just convert that filter to do what you want.

Your [commit](https://github.com/google/shaka-player/commit/e990da01799ff3a1c7355d26ca4e3747b69d3cbe#diff-052c6be0b439274b171a21b6729378b1) to remove the automatic appending of `spc=` didn't go in until ~Aug 5 and there hasn't been a release since before then, so I think this is just something I have to deal with until the next release. That, or the `2.5.4` is shipping the wrong `shaka-player.compiled.js` 🙃 

> This may be #2031. You could try changing the error listener to ignore the error and see if you get playback.

You are suggesting forking shaka player and making the change myself, correct? Or is there an external API of shaka-player's that I can use to turn it off?

> Those are warnings from Shaka Player about invalid configuration paths. The abr configuration should be at the top-level, not part of streaming:

You are correct; I found some setting of the configuration that I hadn't seen before. A quick global search uncovered it.

Thanks again for all of your help; I really appreciate it.

**Comment by TheModMaker** [comment]
> Your commit to remove the automatic appending of spc= didn't go in until ~Aug 5 and there hasn't been a release since before then, so I think this is just something I have to deal with until the next release. That, or the 2.5.4 is shipping the wrong shaka-player.compiled.js upside_down_face

Correct, those changes won't make it into the v2.5.x releases since it's a breaking change.  You'll need to wait until the v2.6 release to see it.  Or you can pull from `master` and build it yourself.

> You are suggesting forking shaka player and making the change myself, correct? Or is there an external API of shaka-player's that I can use to turn it off?

I was talking about forking or just editing the source and build it manually.  I was suggesting a possible solution while we work on fixing #2031.
